File: src/socket-server-linechart.js
// src/live-socket-server.js

import express from 'express';
import http from 'http';
import { Server } from 'socket.io';

// 1. App Express + HTTP server
const app = express();
const server = http.createServer(app);

// 2. Socket.IO with open CORS (dev)
const io = new Server(server, {
  cors: {
    origin: "*",
    methods: ["GET", "POST"]
  }
});

// 3. Baseline trends per cluster (magenta low, orange mid-low, green mid-high)
const clusterTrends = [
  { clusterID: 'Cluster-0', base: 50,  spikeRange: [10, 20] },  // green: starts mid
  { clusterID: 'Cluster-1', base: 30,  spikeRange: [5, 15]  },  // orange: above magenta
  { clusterID: 'Cluster-2', base: 10,  spikeRange: [2, 8]   }   // magenta: always low
];

// 4. Categorie per triplicare le linee
const categories = ['A', 'B', 'C'];

// 5. Initialize the "lines"
const lines = clusterTrends.flatMap(({ clusterID, base, spikeRange }) =>
  categories.map(category => ({
    id: `${clusterID}-${category}`,
    clusterID,
    category,
    base,
    spikeRange,
    y: base + (Math.random() * 6 - 3)  // start jitter
  }))
);

// 6. Stato globale per il contatore x
let globalX = 0;

// 7. Spike probability
const SPIKE_PROBABILITY = 0.05;

// 8. Random helper
function randBetween(min, max) {
  return Math.random() * (max - min) + min;
}

// 9. Linear decay function (from 1 down to minDecay)
function decayFactor(x, startDecayX = 200, minDecay = 0.2) {
  const f = 1 - Math.min(x / startDecayX, 1) * (1 - minDecay);
  return Math.max(f, minDecay);
}

io.on('connection', socket => {
  console.log(`Client connected: ${socket.id}`);

  const interval = setInterval(() => {
    const xValue = globalX++;

    lines.forEach(line => {
      // Light drift towards base
      const driftStrength = 0.005;
      const drift = (line.base - line.y) * driftStrength;

      // Wide noise to encourage crossings
      const noise = randBetween(-10, 10);

      // Apply drift + noise
      line.y += drift + noise;

      // --- Cluster-specific spike handling ---
      if (line.clusterID === 'Cluster-1') {
        // -- Orange: huge initial spikes (x<5), then gradual decay
        if (xValue < 5) {
          // mega-spike at startup
          const mega = randBetween(100, 150);
          line.y += mega;
        } else if (Math.random() < SPIKE_PROBABILITY) {
          // decayed spike:
          const decay = decayFactor(xValue, 500, 0.1);
          const [minS, maxS] = line.spikeRange;
          const spike = randBetween(minS, maxS) / decay;
          line.y += spike;
        }
      }
      else if (line.clusterID === 'Cluster-2') {
        // -- Magenta: never large spikes, always in the low range
        // use only the normal spikeRange
        if (Math.random() < SPIKE_PROBABILITY * 0.5) {
          const [minS, maxS] = line.spikeRange;
          const spike = randBetween(minS, maxS);
          line.y += spike;
        }
      }
      else if (line.clusterID === 'Cluster-0') {
        // -- Green: regular, with occasional medium spikes that decay then rise again
        if (Math.random() < SPIKE_PROBABILITY * 1.2) {
          // decay+bounce cycle: use a slow sine wave
          const osc = (Math.sin(xValue / 100) + 1) / 2;  // da 0 a 1
          const [minS, maxS] = [20, 60];
          const spike = randBetween(minS, maxS) * osc;
          line.y += spike;
        }
      }

      // Emit point
      const point = {
        x: xValue,
        y: Math.round(line.y),
        clusterID: line.clusterID,
        category: line.category
      };
      socket.emit('data', point);
    });

  }, 500);

  socket.on('disconnect', () => {
    console.log(`Client disconnesso: ${socket.id}`);
    clearInterval(interval);
  });
});

// 10. Start on port 3000
const PORT = process.env.PORT || 3000;
server.listen(PORT, () => {
  console.log(`Socket.IO server in ascolto su http://localhost:${PORT}`);
});

File: demo/static/data/socket-server-linechart.py
from flask import Flask
from flask_socketio import SocketIO
import random
import math
import eventlet

eventlet.monkey_patch()  # required for Socket.IO support with eventlet

app = Flask(__name__)
app.config['SECRET_KEY'] = 'secret!'
socketio = SocketIO(app, cors_allowed_origins='*')

# Cluster trends
cluster_trends = [
    {'clusterID': 'Cluster-0', 'base': 50,  'spikeRange': (10, 20)},
    {'clusterID': 'Cluster-1', 'base': 30,  'spikeRange': (5, 15)},
    {'clusterID': 'Cluster-2', 'base': 10,  'spikeRange': (2, 8)},
]

categories = ['A', 'B', 'C']

# Initialize lines
lines = []
for ct in cluster_trends:
    for cat in categories:
        lines.append({
            'id': f"{ct['clusterID']}-{cat}",
            'clusterID': ct['clusterID'],
            'category': cat,
            'base': ct['base'],
            'spikeRange': ct['spikeRange'],
            'y': ct['base'] + (random.uniform(-3, 3))
        })

global_x = 0
SPIKE_PROBABILITY = 0.05

def rand_between(min_val, max_val):
    return random.uniform(min_val, max_val)

def decay_factor(x, start_decay_x=200, min_decay=0.2):
    f = 1 - min(x / start_decay_x, 1) * (1 - min_decay)
    return max(f, min_decay)

@socketio.on('connect')
def handle_connect():
    global global_x
    print('Client connected')

    def send_data(app):
        global global_x
        with(app.app_context()):
            while True:
                x_value = global_x
                global_x += 1

                points = []
                for line in lines:
                    drift_strength = 0.005
                    drift = (line['base'] - line['y']) * drift_strength
                    noise = rand_between(-10, 10)
                    line['y'] += drift + noise

                    if line['clusterID'] == 'Cluster-1':
                        if x_value < 5:
                            mega = rand_between(100, 150)
                            line['y'] += mega
                        elif random.random() < SPIKE_PROBABILITY:
                            decay = decay_factor(x_value, 500, 0.1)
                            min_s, max_s = line['spikeRange']
                            spike = rand_between(min_s, max_s) / decay
                            line['y'] += spike
                    elif line['clusterID'] == 'Cluster-2':
                        if random.random() < SPIKE_PROBABILITY * 0.5:
                            min_s, max_s = line['spikeRange']
                            spike = rand_between(min_s, max_s)
                            line['y'] += spike
                    elif line['clusterID'] == 'Cluster-0':
                        if random.random() < SPIKE_PROBABILITY * 1.2:
                            osc = (math.sin(x_value / 100) + 1) / 2
                            min_s, max_s = 20, 60
                            spike = rand_between(min_s, max_s) * osc
                            line['y'] += spike

                    point = {
                        'x': x_value,
                        'y': round(line['y']),
                        'clusterID': line['clusterID'],
                        'category': line['category']
                    }
                    points.append(point)

                for point in points:
                    socketio.emit('data', point)
                socketio.sleep(0.5)  # pausa 500 ms

    socketio.start_background_task(send_data, app)

@socketio.on('disconnect')
def handle_disconnect():
    print('Client disconnected')

if __name__ == '__main__':
    socketio.run(app, host='0.0.0.0', port=3000)
